sdks/java/harness/src/main/java/org/apache/beam/fn/harness/Caches.java

```diff
@@ -165,10 +141,57 @@ public int weigh(Object key, Object value) {
   }
 
   @VisibleForTesting
-  static Cache<Object, Object> forCache(
-      org.apache.beam.vendor.guava.v26_0_jre.com.google.common.cache.Cache<CompositeKey, Object>
-          cache) {
-    return new SubCache<>(cache, CompositeKeyPrefix.ROOT);
+  static <K, V> Cache<K, V> forMaximumBytes(long maximumBytes) {
+    // We specifically use Guava cache since it allows for recursive computeIfAbsent calls
+    // preventing deadlock from occurring when a loading function mutates the underlying cache
+    LongAdder weightInBytes = new LongAdder();
+    return new SubCache<>(
+        new ShrinkOnEviction(
+                CacheBuilder.newBuilder()
+                    .maximumWeight(maximumBytes >> WEIGHT_RATIO)
+                    .weigher(
+                        new Weigher<Object, WeightedValue<Object>>() {
+
+                          @Override
+                          public int weigh(Object key, WeightedValue<Object> value) {
+                            // Round up to the next closest multiple of WEIGHT_RATIO
+                            long size = ((value.getWeight() - 1) >> WEIGHT_RATIO) + 1;
```

Comment:
It seems like we used to have the cache weigh the key with Caches.weigh(key) but I no longer see that happening.  Was that intentional?

If not, add a test where there are large keys to verify that they are weighed.

Reply:
Originally I was having trouble getting a good way of not paying the cost of weighing the key over and over but I solved that and added it back.

Reply:
I thought it was going to be a hassle to add when I was re-implementing it so we wouldn't have to weigh the key again when removing the entry but it turned out easier then I thought.